File: gatling-core-java/src/main/java/io/gatling/javaapi/core/CounterBuilder.java
/*
 * Copyright 2011-2026 GatlingCorp (https://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.gatling.javaapi.core;

import static io.gatling.javaapi.core.internal.Expressions.*;

import java.util.function.Function;
import org.jspecify.annotations.NonNull;

/**
 * Builder of an action that stores an incrementing value into the virtual users' Session.
 *
 * <p>Immutable, so all methods return a new occurrence and leave the original unmodified.
 */
public final class CounterBuilder implements ActionBuilder {

  private final io.gatling.core.action.builder.CounterBuilder wrapped;

  public CounterBuilder(@NonNull String key) {
    this(io.gatling.core.action.builder.CounterBuilder$.MODULE$.apply(key));
  }

  private CounterBuilder(io.gatling.core.action.builder.CounterBuilder wrapped) {
    this.wrapped = wrapped;
  }

  /**
   * Set the first value to be emitted. Must be positive. Default is 0.
   *
   * @param start the first value
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder startingAt(int start) {
    return new CounterBuilder(wrapped.startingAt(toStaticValueExpression(start)));
  }

  /**
   * Set the first value to be emitted. Must be positive. Default is 0. Only supported when {@link
   * #perUser()} is used, otherwise building the Simulation will fail.
   *
   * @param start the first value, as a Gatling EL String
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder startingAt(@NonNull String start) {
    return new CounterBuilder(wrapped.startingAt(toIntExpression(start)));
  }

  /**
   * Set the first value to be emitted. Must be positive. Default is 0. Only supported when {@link
   * #perUser()} is used, otherwise building the Simulation will fail.
   *
   * @param start the first value, as a function
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder startingAt(@NonNull Function<Session, Integer> start) {
    return new CounterBuilder(wrapped.startingAt(javaIntegerFunctionToExpression(start)));
  }

  /**
   * Set the gap between 2 successive values. Default is 1.
   *
   * @param increment the gap between 2 successive values
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder withIncrement(int increment) {
    return new CounterBuilder(wrapped.withIncrement(toStaticValueExpression(increment)));
  }

  /**
   * Set the gap between 2 successive values. Default is 1. Only supported when {@link #perUser()}
   * is used, otherwise building the Simulation will fail.
   *
   * @param increment the gap between 2 successive values, as a Gatling EL String
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder withIncrement(@NonNull String increment) {
    return new CounterBuilder(wrapped.withIncrement(toIntExpression(increment)));
  }

  /**
   * Set the gap between 2 successive values. Default is 1. Only supported when {@link #perUser()}
   * is used, otherwise building the Simulation will fail.
   *
   * @param increment the gap between 2 successive values, as a function
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder withIncrement(@NonNull Function<Session, Integer> increment) {
    return new CounterBuilder(wrapped.withIncrement(javaIntegerFunctionToExpression(increment)));
  }

  /**
   * Set the inclusive upper bound. Default is Integer.MAX_VALUE. Once it's reached, the load
   * generator is stopped, unless {@link #wrapAround()} is used.
   *
   * @param end the inclusive upper bound
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder upTo(int end) {
    return new CounterBuilder(wrapped.upTo(toStaticValueExpression(end)));
  }

  /**
   * Set the inclusive upper bound. Default is Integer.MAX_VALUE. Once it's reached, the load
   * generator is stopped, unless {@link #wrapAround()} is used. Only supported when {@link
   * #perUser()} is used, otherwise building the Simulation will fail.
   *
   * @param end the inclusive upper bound, as a Gatling EL String
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder upTo(@NonNull String end) {
    return new CounterBuilder(wrapped.upTo(toIntExpression(end)));
  }

  /**
   * Set the inclusive upper bound. Default is Integer.MAX_VALUE. Once it's reached, the load
   * generator is stopped, unless {@link #wrapAround()} is used. Only supported when {@link
   * #perUser()} is used, otherwise building the Simulation will fail.
   *
   * @param end the inclusive upper bound, as a function
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder upTo(@NonNull Function<Session, Integer> end) {
    return new CounterBuilder(wrapped.upTo(javaIntegerFunctionToExpression(end)));
  }

  /**
   * Start over from the first value once the upper bound is reached, instead of stopping the load
   * generator. Beware values are then no longer unique.
   *
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder wrapAround() {
    return new CounterBuilder(wrapped.wrapAround());
  }

  /**
   * Track the counter independently for each virtual user, so they all get the very same sequence
   * of values, instead of sharing one single sequence.
   *
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder perUser() {
    return new CounterBuilder(wrapped.perUser());
  }

  /**
   * Distribute the values evenly amongst all the load generators of a Gatling Enterprise cluster,
   * so they remain unique cluster wide. Only effective when the test is running with Gatling
   * Enterprise, noop otherwise. Each load generator only gets a slice of the range, so the values
   * emitted by the whole cluster have holes.
   *
   * @return a new CounterBuilder
   */
  public @NonNull CounterBuilder shard() {
    return new CounterBuilder(wrapped.shard());
  }

  @Override
  public io.gatling.core.action.builder.ActionBuilder asScala() {
    return wrapped;
  }
}

File: gatling-core/src/main/scala/io/gatling/core/action/builder/CounterBuilder.scala
/*
 * Copyright 2011-2026 GatlingCorp (https://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.gatling.core.action.builder

import java.util.concurrent.atomic.AtomicInteger

import io.gatling.core.action.{ Action, Counter }
import io.gatling.core.session.{ Expression, StaticValueExpression }
import io.gatling.core.structure.ScenarioContext

private[gatling] object CounterBuilder {
  def apply(key: String): CounterBuilder =
    new CounterBuilder(
      key,
      start = StaticValueExpression(0),
      increment = StaticValueExpression(1),
      end = StaticValueExpression(Int.MaxValue),
      wrap = false,
      tracksPerUser = false,
      sharded = false
    )
}

private[gatling] final class CounterBuilder(
    key: String,
    start: Expression[Int],
    increment: Expression[Int],
    end: Expression[Int],
    wrap: Boolean,
    tracksPerUser: Boolean,
    sharded: Boolean
) extends ActionBuilder {
  // shared amongst all the Actions built from this very builder instance, so that using the same builder
  // in multiple places of a Simulation gives one single sequence of values, just like feeders
  private lazy val index = new AtomicInteger

  /**
   * Set the first value to be emitted. Must be positive. Default is 0. Can only be dynamic, eg a function or a Gatling EL String, when [[perUser]] is used.
   */
  def startingAt(newStart: Expression[Int]): CounterBuilder = new CounterBuilder(key, newStart, increment, end, wrap, tracksPerUser, sharded)

  /**
   * Set the gap between 2 successive values. Default is 1. Can only be dynamic, eg a function or a Gatling EL String, when [[perUser]] is used.
   */
  def withIncrement(newIncrement: Expression[Int]): CounterBuilder = new CounterBuilder(key, start, newIncrement, end, wrap, tracksPerUser, sharded)

  /**
   * Set the inclusive upper bound. Default is Int.MaxValue. Once it's reached, the load generator is stopped, unless [[wrapAround]] is used. Can only be
   * dynamic, eg a function or a Gatling EL String, when [[perUser]] is used.
   */
  def upTo(newEnd: Expression[Int]): CounterBuilder = new CounterBuilder(key, start, increment, newEnd, wrap, tracksPerUser, sharded)

  /**
   * Start over from the first value once the upper bound is reached, instead of stopping the load generator. Beware values are then no longer unique.
   */
  def wrapAround: CounterBuilder = new CounterBuilder(key, start, increment, end, wrap = true, tracksPerUser, sharded)

  /**
   * Track the counter independently for each virtual user, so they all get the very same sequence of values, instead of sharing one single sequence.
   */
  def perUser: CounterBuilder = new CounterBuilder(key, start, increment, end, wrap, tracksPerUser = true, sharded)

  /**
   * Distribute the values evenly amongst all the load generators of a Gatling Enterprise cluster, so they remain unique cluster wide. Only effective when the
   * test is running with Gatling Enterprise, noop otherwise. Each load generator only gets a slice of the range, so the values emitted by the whole cluster
   * have holes, which is the price to pay for not having to synchronize the load generators.
   */
  def shard: CounterBuilder = new CounterBuilder(key, start, increment, end, wrap, tracksPerUser, sharded = true)

  override def build(ctx: ScenarioContext, next: Action): Action = {
    require(!(tracksPerUser && sharded), s"Counter '$key' can't be both perUser and shard as per user counters are not shared in the first place")

    (start, increment, end) match {
      case (StaticValueExpression(startValue), StaticValueExpression(incrementValue), StaticValueExpression(endValue)) =>
        Counter.validateRange(key, startValue, incrementValue, endValue).onFailure(message => throw new IllegalArgumentException(message))

        val count = Counter.valueCount(startValue, incrementValue, endValue)

        val (first: Int, length: Long) =
          (startValue, count)

        if (tracksPerUser) {
          new Counter.PerUser(key, first, incrementValue, length, wrap, ctx.coreComponents.controller, ctx.coreComponents.statsEngine, next)
        } else {
          new Counter.Shared(key, first, incrementValue, length, wrap, index, ctx.coreComponents.controller, ctx.coreComponents.statsEngine, next)
        }

      case _ =>
        require(
          tracksPerUser,
          s"Counter '$key' can only use dynamic startingAt, withIncrement and upTo values, such as functions and Gatling EL Strings, when perUser is used"
        )

        new Counter.PerUserDynamic(key, start, increment, end, wrap, ctx.coreComponents.controller, ctx.coreComponents.statsEngine, next)
    }
  }
}
